pkg/gds/store/trtl/iterator.go

```diff
@@ -0,0 +1,222 @@
+package trtl
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+
+	"github.com/rs/zerolog/log"
+	trtlpb "github.com/trisacrypto/directory/pkg/trtl/pb/v1"
+	"github.com/trisacrypto/directory/pkg/utils/wire"
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+	"google.golang.org/protobuf/proto"
+)
+
+type iterWrapper struct {
+	iter *trtlIterator
+}
+
+type vaspIterator struct {
+	iterWrapper
+}
+
+func (i *iterWrapper) Next() bool {
+	return i.iter.Next()
+}
+
+func (i *iterWrapper) Prev() bool {
+	return i.iter.Prev()
+}
+
+func (i *iterWrapper) Error() error {
+	return i.iter.Error()
+}
+
+func (i *iterWrapper) Release() {
+	i.iter.Release()
+}
+
+type trtlIterator struct {
+	client        trtlpb.TrtlClient
+	values        []*trtlpb.KVPair
+	index         int
+	nextPageToken string
+	cursor        trtlpb.Trtl_CursorClient
+	cancel        context.CancelFunc
+	snapshot      bool
+	namespace     string
+	err           error
+}
+
+func NewTrtlIterator(client trtlpb.TrtlClient, snapshot bool, namespace string) (iter *trtlIterator) {
+	iter = &trtlIterator{
+		client:    client,
+		snapshot:  snapshot,
+		namespace: namespace,
+	}
+
+	var ctx context.Context
+	ctx, iter.cancel = getContext()
+
+	if snapshot {
+		request := &trtlpb.CursorRequest{
+			Namespace: namespace,
+		}
+		iter.cursor, iter.err = iter.client.Cursor(ctx, request)
+		iter.values = make([]*trtlpb.KVPair, 0)
+	} else {
+		defer iter.cancel()
+		request := &trtlpb.IterRequest{
+			Namespace: namespace,
+		}
+
+		var reply *trtlpb.IterReply
+		var err error
+		if reply, err = iter.client.Iter(ctx, request); err != nil {
+			iter.err = err
+			return
+		}
+		iter.values = reply.Values
+		iter.nextPageToken = reply.NextPageToken
+	}
```

Comment:
I recommend making the first request on `Next` instead of `New` so that the client can perform other actions (like Seek) before calling `Next`).